// browser-pool/dist/abstract-classes/browser-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BrowserPlugin = void 0;
const tslib_1 = require("tslib");
const lodash_merge_1 = tslib_1.__importDefault(require("lodash.merge"));
const launch_context_1 = require("../launch-context");
const utils_1 = require("./utils");
;
/**
 * The `BrowserPlugin` serves two purposes. First, it is the base class that
 * specialized controllers like `PuppeteerPlugin` or `PlaywrightPlugin` extend.
 * Second, it allows the user to configure the automation libraries and
 * feed them to {@link BrowserPool} for use.
 */
class BrowserPlugin {
    constructor(library, options = {}) {
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.constructor.name
        });
        Object.defineProperty(this, "library", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "launchOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "proxyUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "userDataDir", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "useIncognitoPages", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        const { launchOptions = {}, proxyUrl, userDataDir, useIncognitoPages = false, } = options;
        this.library = library;
        this.launchOptions = launchOptions;
        this.proxyUrl = proxyUrl && new URL(proxyUrl).href;
        this.userDataDir = userDataDir;
        this.useIncognitoPages = useIncognitoPages;
    }
    /**
     * Creates a `LaunchContext` with all the information needed
     * to launch a browser. Aside from library specific launch options,
     * it also includes internal properties used by `BrowserPool` for
     * management of the pool and extra features.
     */
    createLaunchContext(options = {}) {
        const { id, launchOptions = {}, proxyUrl = this.proxyUrl, useIncognitoPages = this.useIncognitoPages, userDataDir = this.userDataDir, } = options;
        return new launch_context_1.LaunchContext({
            id,
            launchOptions: (0, lodash_merge_1.default)({}, this.launchOptions, launchOptions),
            browserPlugin: this,
            proxyUrl,
            useIncognitoPages,
            userDataDir,
        });
    }
    createController() {
        return this._createController();
    }
    /**
     * Launches the browser using provided launch context.
     */
    async launch(launchContext = this.createLaunchContext()) {
        const { proxyUrl, launchOptions } = launchContext;
        if (proxyUrl) {
            await this._addProxyToLaunchOptions(launchContext);
        }
        if (this._isChromiumBasedBrowser(launchContext)) {
            // This will set the args for chromium based browsers to hide the webdriver.
            launchOptions.args = this._mergeArgsToHideWebdriver(launchOptions.args);
        }
        return this._launch(launchContext);
    }
    _mergeArgsToHideWebdriver(originalArgs) {
        if (!(originalArgs === null || originalArgs === void 0 ? void 0 : originalArgs.length)) {
            return ['--disable-blink-features=AutomationControlled'];
        }
        const argumentIndex = originalArgs.findIndex((arg) => arg.startsWith('--disable-blink-features='));
        if (argumentIndex !== -1) {
            originalArgs[argumentIndex] += ',AutomationControlled';
        }
        else {
            originalArgs.push('--disable-blink-features=AutomationControlled');
        }
        return originalArgs;
    }
    ;
    /**
     * @private
     */
    // @ts-expect-error Give runtime error as well as compile time
    // eslint-disable-next-line space-before-function-paren, @typescript-eslint/no-unused-vars, max-len
    _addProxyToLaunchOptions(launchContext) {
        (0, utils_1.throwImplementationNeeded)('_addProxyToLaunchOptions');
    }
    // @ts-expect-error Give runtime error as well as compile time
    // eslint-disable-next-line space-before-function-paren, @typescript-eslint/no-unused-vars, max-len
    _isChromiumBasedBrowser(launchContext) {
        (0, utils_1.throwImplementationNeeded)('_isChromiumBasedBrowser');
    }
    /**
     * @private
     */
    // @ts-expect-error Give runtime error as well as compile time
    // eslint-disable-next-line space-before-function-paren, @typescript-eslint/no-unused-vars, max-len
    _launch(launchContext) {
        (0, utils_1.throwImplementationNeeded)('_launch');
    }
    /**
     * @private
     */
    // @ts-expect-error Give runtime error as well as compile time
    // eslint-disable-next-line space-before-function-paren
    _createController() {
        (0, utils_1.throwImplementationNeeded)('_createController');
    }
}
exports.BrowserPlugin = BrowserPlugin;
